File: normal_order_python/234.PalindromeLinkedList.py
'''

234. Palindrome Linked List

Easy

Given a singly linked list, determine if it is a palindrome.

Example 1:

Input: 1->2
Output: false
Example 2:

Input: 1->2->2->1
Output: true
Follow up:
Could you do it in O(n) time and O(1) space?

'''

class Solution:
	# O(L) time, O(n) extra space
    def isPalindrome(self, head: ListNode) -> bool:
        check = []
        if not head:
            return True
        while head:
            check.append(head.val)
            head = head.next
        return check == check[::-1]

    # O(L) time, O(1) extra space
    def isPalindrome(self, head: ListNode) -> bool:
    slow = fast = head
    reverse = None
    while fast and fast.next:
        reverse, reverse.next, slow, fast = slow, reverse, slow.next, fast.next.next
    if fast:
        slow = slow.next
    while slow:
        if slow.val == reverse.val:
            slow, reverse = slow.next, reverse.next
        else:
            return False
    return True

File: normal_order_java/234.PalindromeLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

class Solution {
    public boolean isPalindrome(ListNode head) {
        List<ListNode> ll = new ArrayList<ListNode>();
        while (head!=null){
            ll.add(head);
            head = head.next;
        }
        ListNode a[] = ll.toArray(new ListNode[ll.size()]);
        // System.out.println(a);
        int j = a.length - 1;
        for(int i = 0; i < a.length; i++){
            if (a[i].val != a[j].val){
                return false;
            }
            j--;
        }
        return true;
    }
}

class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode l1 = head;
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        
        // reverse 2nd half linked list
        ListNode rev = null;
        while (slow != null){
            ListNode tmp = rev;
            rev = slow;
            slow = slow.next;
            rev.next = tmp;
        }

        // check palindrome
        while (rev != null){
            if (rev.val != l1.val){
                return false;
            }
            rev = rev.next;
            l1 = l1.next;
        }
        return true;
        
    }
}